### PR TITLE
Flux delete sync event

### PR DESCRIPTION
**Pull request by latchmihay**
Flux supports the deletion of objects, its just like a regular sync event but it has no serviceID's, the existing code was strictly checking for serviceID's and slack messages were dependent on serviceID's as well. This commit is to add no serviceID support (aka, flux delete sync event)

### CODE REVIEW

**Comment by fdpeiter** [comment]
@latchmihay we've updated the go module for flux to the 1.19.0.
Could you please rebase and update any legacy reference?

File: pkg/exporters/slack_test.go [hunk 1]
```diff
@@ -126,6 +126,29 @@ func TestNewSlackMessage(t *testing.T) {
 	assert.Equal(t, message.Title, attach.Title)
 }
 
+func TestNewSlackMessageNoIDs(t *testing.T) {
+	message := msg.Message{
+		TitleLink: "https://myvcslink/",
+		Title:     "The title of the message",
+		Body:      "this is the message body",
+		Event: fluxevent.Event{
+			ServiceIDs: []flux.ResourceID{},
```
Comment (fdpeiter) [inline pkg/exporters/slack_test.go]: After rebasing, I think you need to update this reference
```suggestion
			ServiceIDs: []resource.ID{},
```

**Review by fdpeiter — CHANGES_REQUESTED** [review]

**Review by fdpeiter — APPROVED** [review]

**Comment by fdpeiter** [comment]
You can get the docker image with:
`docker pull quay.io/tfgco/fluxcloud`

File: pkg/exporters/slack.go [hunk 2]
```diff
@@ -171,6 +170,16 @@ func (s *Slack) parseSlackChannelConfig(channels string) error {
 // Match namespaces from service IDs to Slack channels.
 func (s *Slack) determineChannels(message msg.Message) []string {
 	var channels []string
+	// When flux delete objects there are no ServiceIDs
```
Comment (marcosnils) [inline pkg/exporters/slack.go:173]: I guess this function could be refactored?? IIUC we could have a single for loop to determine the slack channels?

File: pkg/formatters/default.go [hunk 3]
```diff
@@ -129,10 +129,6 @@ func NewDefaultFormatter(config config.Config) (*DefaultFormatter, error) {
 
 // Format plaintext message for an exporter for Flux event
 func (d DefaultFormatter) FormatEvent(event fluxevent.Event, exporter exporters.Exporter) msg.Message {
-	if len(event.ServiceIDs) == 0 {
```
Comment (marcosnils) [inline pkg/formatters/default.go:132]: Interesting that this got removed and no tests had to be changed :thinking:

**Comment by dan-slinky-ckpd** [comment]
@latchmihay Thanks for this PR, we've been watching and just took the opportunity to test `0.4.0` on an environment with garbage collection enabled, however fluxcloud didn't receive an event for sync delete. Do you have it running successfully somewhere?

**Comment by dan-slinky-ckpd** [comment]
Actually, an event with empty serviceID was received, but it didn't send a Slack message.

```
{"id":0,"serviceIDs":[],"type":"sync","startedAt":"2020-07-06T16:39:42.068985084Z","endedAt":"2020-07-06T16:39:42.068985084Z","logLevel":"info","metadata":{"commits":[{"revision":"32202e7dc85f51eae2261806059bb031162b8ada","message":"remove thing flux patch (#568)"},{"revision":"9fbe60286e2cbee2a0e26ff2f68c5331e7617a7e","message":"disable thing (#567)"}],"includes":{"other":true}}}
```

We're running `fluxcd/flux-prerelease:master-c26e9fe8`

**Comment by latchmihay** [comment]
> @latchmihay Thanks for this PR, we've been watching and just took the opportunity to test `0.4.0` on an environment with garbage collection enabled, however fluxcloud didn't receive an event for sync delete. Do you have it running successfully somewhere?

No I don't, I am not sure if something has changed in Flux 1.19 in terms of the delete payload. Take a look at the tests provided and the payload on a delete event I am basing this on

**Comment by dan-slinky-ckpd** [comment]
@latchmihay Ignore my comments above, your changes is correct and working as expected. The change I'm waiting for is https://github.com/fluxcd/flux/pull/3127 which will then just require a small update to the default formatter to include a list of resources. Thanks again for this contribution!

**Comment by fdpeiter** [comment]
Hey @dan-slinky-ckpd! We haven't checked [this](https://github.com/fluxcd/flux/issues/2494#issuecomment-579359175) suggested change for the formatter yet, but I think that will fix this problem, right?

**Comment by dan-slinky-ckpd** [comment]
@fdpeiter I'm not sure you need to make that change as the spurious list of resources issue has been resolved by https://github.com/fluxcd/flux/pull/3039. The change I'm referring to above is once https://github.com/fluxcd/flux/pull/3127 has been merged and a new flux release is published we can update the formatter to include `event.DeletedWorkloadIDs`, [for example](https://github.com/dan-slinky-ckpd/fluxcloud/commit/44ba530fb040a8beaf28a9f52b4204f0ede2e778#diff-8eabebfe03669fc383bd0098dd20fd49).

The result of that example looks like this:

Screenshot 2020-07-07 at 16 59 22

If a user does want to adapt the message to exclude the list of resources completely they should be able to override the template with `BODY_TEMPLATE`, but I've not tried that myself.

**Comment by fdpeiter** [comment]
Got it, gonna keep an eye on it to add this whenever merged.

**Pull request merged**